### PR TITLE
Fix Unparser compatibility

### PR DESCRIPTION
**Pull request by palkan**
Unparser 0.5.0 doesn't work as we expect, let's figure this out...

### CODE REVIEW

**Comment by palkan** [comment]
@mbj Look what I found 😀

```
 ruby -runparser -e '
puts Unparser.unparse(Unparser.parse("[%(), %q(), %Q()]"))
'

[<<-HEREDOC, <<-HEREDOC, <<-HEREDOC]
HEREDOC
HEREDOC
HEREDOC
````

Original failing example is this one: https://github.com/ruby/ruby/blob/501f2c44e6ae79c02a5c4d0f872fc7fa77258fcf/test/ruby/test_pattern_matching.rb#L350-L355

**Comment by mbj** [comment]
@palkan you found a way to create empty `dstr` nodes. I can make unparser emit more "beauyful" source for these now.

Before I was under the assumption there is no way to generate empty `dstr` nodes without an empty heredoc.

But, the unparse guarantee actually is uphold right now (on non nice code arguably):

```
bundle exec unparser -e '[%(), %q(), %Q()]' --verbose                 <<<
(string)
Original-Source:
[%(), %q(), %Q()]
Generated-Source:
[<<-HEREDOC, <<-HEREDOC, <<-HEREDOC]
HEREDOC
HEREDOC
HEREDOC
Original-Node:
(array
  (dstr)
  (dstr)
  (dstr))
Generated-Node:
(array
  (dstr)
  (dstr)
  (dstr))
Success: (string)
```

So unparser "does uphold" its guarantee here, it just produces non minimal concrete syntax.

**Comment by mbj** [comment]
Just to make sure we are on the same page: I'm going to change `(dstr)` to be emitted as `%()` now.

**Comment by mbj** [comment]
See: https://github.com/mbj/unparser/pull/167

**Comment by mbj** [comment]
@palkan A syntax generation change was released as 0.5.1. Note that conceptually the original syntax was also "right" (As it produces the same AST nodes). If that syntax problem caused for you it means that your integration to unparser has a problem.

There are strings unparser has to emit as heredocs, as the ASTs can only be parsed from a heredoc. And if the generated heredoc syntax caused a problem for your case, than it'll still in other scenarios.

File: bin/reparse [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+contents =
+  if File.exist?(ARGV[0])
+    File.read(ARGV[0])
+  else
+    ARGV[0]
+  end
+
+ast = RubyNext::Language.parse(contents)
+puts Unparser.unparse(ast)
```
Comment (mbj) [inline bin/reparse:19]: You could reuse the `Unparser::Verification` class here to print more detials equivalent to the ones `unparser` CLI will.

Comment (palkan) [reply bin/reparse:19]: I'm using a custom parser, and, as far as I can see, Unparser has its own.

Thus, I'm using this script to verify that Unparser correctly works with my parser.

Comment (mbj) [reply bin/reparse:19]: its easy to change `Unparser::Verification` to accept a custom parser. That class produces a nicer report.

**Comment by palkan** [comment]
@mbj Found a regression in Ruby 2.3 support: https://github.com/ruby-next/ruby-next/pull/61/checks?check_run_id=1235633688

**Comment by mbj** [comment]
> Shouldn't there be a `required_ruby_version` in the gemspec?

I've had people explicitly ask me to remove these version checks in the gemspec, as they wanted to install unpraser and mutant in CI matrixes without having to maintain per matrix cell gemfiles.

Unparser targets what mutant is targeting: https://github.com/mbj/mutant#ruby-versions

**Comment by mbj** [comment]
@palkan I'm happy to support lower ruby versions once I get demand from it from mutant customers that drive unparser evolution. But so far all of them are on >= 2.5.

**Comment by mbj** [comment]
@palkan Yeah so it appeared you had been programming against unparser bugs before. Now, given unparsers invariant is strongly enforced (without exceptions) unparser releases requirering changes on your side should not happen as often anymore.

**Pull request merged**